### PR TITLE
initial vector implementation

### PR DESCRIPTION
**Pull request by drakemain**
Can construct vector from initializer_list. Can push() and pop(). @blaenk

### CODE REVIEW

File: c++/build.sh [hunk 1]
```diff
@@ -1,3 +1,4 @@
 #!/usr/bin/env sh
 clang++ -std=c++11 -Wall -Wextra -Werror -g ./src/linked-list.cc -o out/linked-list.bin
-clang++ -std=c++11 -Wall -Wextra -Werror -g ./src/binary-tree.cc -o out/binary-tree.bin
\ No newline at end of file
+clang++ -std=c++11 -Wall -Wextra -Werror -g ./src/binary-tree.cc -o out/binary-tree.bin
+clang++ -std=c++11 -Wall -Wextra -Werror -g ./src/vector.cc -o out/vector.bin
```
Comment (blaenk) [inline c++/build.sh:4]: Just as a note. As you know extensions don't matter, but typically binaries don't have an extension.

File: c++/src/vector.cc [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+
+
+  void push(T val) {
+    if (length >= capacity) {
+      std::cout << "Vector capacity exceeded. ";
+      std::exit(1);
+    }
+
+    this->dataStore[this->length] = val;
+    this->length++;
```
Comment (blaenk) [inline c++/src/vector.cc]: This is perfectly fine. Another way you can do it is to reduce those two lines into one.

``` cpp
this->dataStore[this->length++] = val;
```

It's not clearly better. I think I would keep what you have, though I would change the second line to:

``` cpp
++this->length;
```

Because it's potentially more efficient. In general use pre-increment unless you know you need post-increment.

File: c++/src/vector.cc [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+  }
+
+  T pop() {
+    if (length == 0) {
+      std::cout << "Attempted to pop empty vector. ";
+      std::exit(1);
+    }
+
+    length--;
+    return this->dataStore[length + 1];
```
Comment (blaenk) [inline c++/src/vector.cc]: This reads very weird. You subtract one but then you add one cause you wanted the previous value, but you want it to stay subtracted. This feels like a perfect opportunity for post-decrement, which gives you the current value for the place you're using it, but ensures that the value is decremented by the end of the statement.

``` cpp
return this->dataStore[this->length--];
```

You can think of that as expanding more or less to:

``` cpp
T val = this->dataStore[length];
length--;
return val;
```

Which I think would look better than what you have. The reason what you have is confusing (relative to the other options) is that you change length then you change it back for a temporary reason.

File: c++/src/vector.cc [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+      std::cout << "Vector capacity exceeded. ";
+      std::exit(1);
+    }
+
+    this->dataStore[this->length] = val;
+    this->length++;
+  }
+
+  T pop() {
+    if (length == 0) {
```
Comment (blaenk) [inline c++/src/vector.cc]: `this->length`

Remember, if you're gonna use `this->`, use it everywhere.

File: c++/src/vector.cc [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+    }
+  }
+
+  ~Vector() {
+    delete[] this->dataStore;
+  }
+
+
+  void push(T val) {
+    if (length >= capacity) {
```
Comment (blaenk) [inline c++/src/vector.cc]: `this->length`, `this->capacity`, etc.

File: c++/src/vector.cc [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    length--;
+    return this->dataStore[length + 1];
+  }
+
+  void toOstream(std::ostream &stream) {
+    for (unsigned int i = 0; i < this->length; i++) {
+      stream << this->dataStore[i];
+
+      if (i < this->length - 1) {
+        stream << " → ";
```
Comment (blaenk) [inline c++/src/vector.cc]: This is fine btw, but I used that arrow myself for linked lists or anything that links. For arrays its more conventional to print them as you type them in code, i.e. `[1, 2, 3]`. It doesn't matter here, but just as a note.

File: c++/src/vector.cc [hunk 7]
```diff
@@ -0,0 +1,100 @@
...
+      }
+    }
+  }
+};
+
+template <typename T>
+void test_expected_values(Vector<T> &vector, std::string expectedString);
+
+void test_push() {
```
Comment (blaenk) [inline c++/src/vector.cc]: Good work here. However, it's preferable to try to keep one "test" per test. So right here you have two. The idea you want to aim for is that if a test crashes, it's for one reason and one reason only, so you right away know what's wrong. Here there are two tests. If they're testing the same thing, i.e. they would crash for the same reason, then it's just wasted time on your part to write two: they're redundant, and they don't _technically_ make things safer, they just make you feel more at ease, but _technically_/_practically_ it's not necessary, kind of like earlier how "just in case I'll use `>=`", except worse here cause it entails unnecessary work.

So instead of a very broad `test_push` I would create `test_push_when_full()`, `test_push_when_empty()`, etc. if you want. You can still keep `test_push` as a way of grouping the push tests if you want, so in `test_push` you would call `test_push_when_full` et al, then in main you just call `test_push`. The benefit of this is that if you only want to focus on pop tests for whatever reason, you would be able to just comment out `test_push()` and leave `test_pop()`.

File: c++/src/vector.cc [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+  Vector<int> push_vector{1, 2, 3};
+  push_vector.push(4);
+  test_expected_values(push_vector, "1 → 2 → 3 → 4");
+  push_vector.push(1234);
+  test_expected_values(push_vector, "1 → 2 → 3 → 4 → 1234");
```
Comment (blaenk) [inline c++/src/vector.cc]: I'm not quite sure what this is testing. I guess it's just a regular-use test. You can call it maybe `test_normal_push()`.

File: c++/src/vector.cc [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+
+  // test on empty vector
+  Vector<int> push_vector_empty{};
+  push_vector_empty.push(1);
+  test_expected_values(push_vector_empty, "1");
```
Comment (blaenk) [inline c++/src/vector.cc]: This you can move into a `test_push_when_empty()`. BTW you can use camel case too, I was using snake case cause I'm too used to Rust. Either one, doesn't matter. Personally I think it's sometimes easier to read for tests instead of `testPushWhenEmpty` all squished together, but it doesn't really matter.

File: c++/src/vector.cc [hunk 10]
```diff
@@ -0,0 +1,100 @@
...
+
+  std::cout << "Push tests passed." << std::endl;
+}
+
+void test_pop() {
+  Vector<int> pop_vector{1, 2, 3, 4, 5};
+  pop_vector.pop();
+  test_expected_values(pop_vector, "1 → 2 → 3 → 4");
```
Comment (blaenk) [inline c++/src/vector.cc]: Perhaps make this `test_normal_pop()`. Personally I would make this test simpler and just put one element in it, then pop, and test that the vector is now empty. Notice how this would necessitate a kind of `.length()` method which you can easily implement and would be useful.

Notice how tests drive the implementation of your code. Perhaps another test you create might want to be with respect to the capacity, later on, for example you want to ensure that when you cause the vector to grow, it grows to double the capacity, so you would save the previous capacity via some `.capacity()` method, push to cause a grow, then test that `previousCapacity * 2 == vector.capacity()`, i.e. that it doubled in capacity. A consequence of letting your tests drive your implementation, i.e. you only implement things that your tests need, means that it's much more likely that all code you write is properly tested.

_Don't_ rush to implement `capacity()` right now. The point is that you can let your tests drive your implementation (i.e. TDD). Otherwise you'd be creating unnecessary work right now cause you don't need it ATM.

Basically try to keep the tests as minimal as you need to to make the test work. This way there are less pieces to figure out if something goes wrong. You can extend this into more complex code/tests and see how this would be a big win.

File: c++/src/vector.cc [hunk 11]
```diff
@@ -0,0 +1,100 @@
...
+  std::cout << std::endl;
+
+  test_push();
+  test_pop();
+
+  std::cout << std:: endl << "All tests passed" << std::endl;
+};
+
+template <typename T>
+void test_expected_values(Vector<T> &vector, std::string expectedString) {
```
Comment (blaenk) [inline c++/src/vector.cc]: `const Vector<T> &vector`, so that you communicate that you don't modify the vector in this function, and to _ensure_ that you don't (barring some escape hatches that you won't use).

You can also do `const std::string &expectedString`.

**Comment by blaenk** [comment]
Overall good work. Made extensive comments.

By the way, one thing I think you should do instead of doing `exit()` now that I think about it is you should throw an exception.

Include `stdexcept` then you can `throw std::runtime_error("tried to push on full vector!");`.

Then what you can do is cause it to throw an exception and you can test that it did indeed throw an exception.

``` cpp
void test_push_when_full_throws_exception() {
  Vector<int> vec{1, 2, 3, 4, 5, 6, 7, 8}; // i.e. it's full cause 8 slots
  bool threw = false;

  try {
    vec.push(9);
  } catch (...) {
    threw = true;
  }

  assert(threw);
}
```

You should do this for sure. Otherwise you're currently not testing that branch/case. It's straightforward, though I did this by hand so I'm not sure if it's exactly correct.

http://en.cppreference.com/w/cpp/language/try_catch

File: c++/build.sh (hunk 1 above)
Comment (blaenk) [reply c++/build.sh:4]: Also it might be worth to start looking into either makefiles or cmake (what I use). Here's a [cmake guide](https://cgold.readthedocs.io/en/latest/first-step/minimal-example.html). CMake is cross platform. Using something like one of these (CMake uses makefiles under the hood on linux) will help with for example only building the stuff that changed etc.

Comment (blaenk) [reply c++/build.sh:4]: Here's a [simple guide](http://mrbook.org/blog/tutorials/make/) on the basics of makefiles if you prefer to go that route.

**Comment by blaenk** [comment]
You can even factor that out into a separate helper function I'll make for you:

``` cpp
template <typename Function>
void assert_throws(Function func) {
  bool threw = false;

  try {
    func();
  } catch (...) {
    threw = true;
  }

  assert(threw);
}
```

I think that should work. Then you can change the above test to:

``` cpp
void test_push_when_full_throws_exception() {
  Vector<int> vec{1, 2, 3, 4, 5, 6, 7, 8};

  assert(vec.length() == vec.capacity()); // assert that it's full
  assert_throws([&] { vec.push(9); });    // assert that pushing throws
}
```

So that asserts/ensures that the code `vec.push(9)` throws an exception.

I did it by hand so I haven't tested it but it should work. The benefit of factoring/DRYing stuff like this is that you can then re-use it in other places. It also makes the tests much more self-describing. You can read the test faster and the combination of the descriptive name, short/concise test code, and the fact that you're only testing one thing, makes the test very understandable.

File: c++/src/vector.cc (hunk 3 above)
Comment (drakemain) [reply c++/src/vector.cc]: Yeah, I didn't like how I did that. Couldn't remember how the post/pre-increment/decrement worked.

File: c++/src/vector.cc (hunk 9 above)
Comment (drakemain) [reply c++/src/vector.cc]: I was thinking of using underscores in the tests and camelcase for non-test functions and methods.

Comment (blaenk) [reply c++/src/vector.cc]: Sounds good to me.

File: c++/src/vector.cc (hunk 3 above)
Comment (blaenk) [reply c++/src/vector.cc]: Fair enough.

File: c++/src/vector.cc [hunk 12]
```diff
@@ -69,43 +69,71 @@ class Vector {
...
-  test_expected_values(push_vector_empty, "1");
+  test_expected_values(push_vector_empty, "[1]");
+}
+
+void test_full_push_throws_exception() {
+  Vector<int> push_vector_full{1, 2, 3, 4, 5, 6, 7, 8};
+  assert(push_vector_full.cap() == push_vector_full.len());
 
-  std::cout << "Push tests passed." << std::endl;
+  test_throws_exception([&] {push_vector_full.push(9); });
```
Comment (blaenk) [inline c++/src/vector.cc:92]: space between opening brace `{` and `p`
